### ISSUE TITLE
torch.Tensor.new_full ignores requires grad

### ISSUE THREAD

**Issue by bobiblazeski** [pytorch/pytorch#36455]
## To Reproduce

1. Create tensor
2. Create new_full with requires_grad=True
3. New full tensor should require gradient as per [documentation](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.new_full)

```
tensor = torch.ones((2,))
new_tensor = tensor.new_full((3, 4), 3.141592, requires_grad=True)
print(new_tensor.requires_grad)
# False
tensor = torch.ones((2,), requires_grad=True)
new_tensor = tensor.new_full((3, 4), 3.141592, requires_grad=True)
print(new_tensor.requires_grad)
# False
new_tensor.requires_grad = True
print(new_tensor.requires_grad)
# True
```

## Environment
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.13.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 440.33.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.4

Versions of relevant libraries:
[pip3] botorch==0.1.2
[pip3] efficientnet-pytorch==0.5.1
[pip3] gpytorch==0.3.4
[pip3] numpy==1.18.0
[pip3] pytorch3d==0.1
[pip3] torch==1.4.0
[pip3] torch-cluster==1.5.4
[pip3] torch-geometric==1.4.3
[pip3] torch-scatter==2.0.4
[pip3] torch-sparse==0.6.1
[pip3] torchcontrib==0.0.2
[pip3] torchdiffeq==0.0.1
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.4
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.5.0
[conda] _pytorch_select           0.2                       gpu_0  
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.0.130                      0  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] msgpack-numpy             0.4.3.2                  py37_0  
[conda] numpy                     1.17.3           py37hd14ec0e_0  
[conda] numpy-base                1.17.3           py37hde5b4d6_0  
[conda] pytorch                   1.2.0           cuda100py37h938c94c_0  
[conda] torchvision               0.4.0           cuda100py37hecfc37a_0

## Additional context

No

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by ezyang** [comment]
This really should work, upgrading priority

**Comment by prajjwal1** [comment]
@ezyang Can I work on this issue ? Also, t would be great if you could provide a starting point. Thanks!

**Comment by ezyang** [comment]
@prajjwal1 Sure. However it may be a bug deep in internals making it difficult to track down. I would probably start by looking at the Python bindings for this function; they live in `torch/csrc/autograd/generated/python_variable_methods.cpp` after you do a build (nb this is a generate file) and then walk down the call stack until you find where requires grad is being clobbered.

**Comment by ezyang** [comment]
Alt, the bug might be that we are not ever hitting the VariableType bindings for new_full

**Comment by prajjwal1** [comment]
Thanks @ezyang. I'll try to look into this and will share my findings (hopefully send a PR).

**Comment by prajjwal1** [comment]
@ezyang, I also looked at `torch.Tensor.new_empty()`:
```
>>> import torch
>>> tensor = torch.ones((), requires_grad=True)
>>> tensor = tensor.new_empty((2,3))
>>> tensor.requires_grad
False
>>> print(tensor.requires_grad)
False
>>> tensor = torch.ones(())
>>> tensor.new_empty((2, 3), requires_grad=True)
tensor([[-6.7072e-02,  4.5623e-41,  4.6277e-37],
        [ 0.0000e+00,  7.9423e+25,  4.5623e-41]])
>>> tensor = tensor.new_empty((2, 3), requires_grad=True)
>>> print(tensor.requires_grad)
False
>>> tensor.requires_grad=True
>>> print(tensor.requires_grad)
True
```
Is this behaviour also intended ? As per the documentation, it should have the same bool value for `requires_grad` as the source tensor.

**Comment by prajjwal1** [comment]
I am new to autograd C++ code. Are changes to be made here ?
```
static PythonArgParser parser({
"new_full(IntArrayRef size , .......... , bool requires_grad=False)",}
I think it should be something like `bool requires_grad=self_.requires_grad` (not sure) if self_ has an attribute `requires_grad`.

**Comment by ezyang** [comment]
Yes, probably all the `new_` functions are broken all the same way.

I don't think the ArgParser is the problem: `bool requires_grad=False` just says that if the argument isn't available, default it to False (which does seem to be working correctly).

**Comment by prajjwal1** [comment]
Thanks for replying.
1.  I noticed that all new functions are returned by
```
wrap(dispatch_new_zeros/full/empty(self, _r.intlist(0), options));
```
while some functions like apply, size are returned by
```
THPVariable_Wrap()
```
I'm not sure of the difference between these two wrap functions. 
Did you mean this when you said ?
> Alt, the bug might be that we are not ever hitting the VariableType bindings for new_full

2. This seems like a generated file (csrc/autograd/generated), where should I make the change so that generated file contains the reflected changes ?

**Comment by ezyang** [comment]
VariableType is generated in tools/autograd/gen_variable_type.py

**Comment by prajjwal1** [comment]
Thanks for the reply. I'm seeing `new_` functions being present inside `torch/csrc/autograd/generated`. I searched all of the tools directory. This `aten/src/ATen/native/TensorFactories.cpp` contains the definition. Where are they defined outside `generated` so that I can make the changes which would be reflected in `generated` after build?

**Comment by albanD** [comment]
As you can see in `generated/python_variable_methods.cpp`, the kwargs of python are properly collected and packed into a `TensorOption`.
This is then calling into the one in `generated/VariableType_4.cpp` that simply transfers the call as is to ATen.
This is handled in `TensorFactories.cpp` that root it to `at::empty()`.
This one you can check in `native_functions.yaml` that it is routed to different functions for different backend. In particular, for cpu, it is `empty_cpu()` that you can find in `TensorFactories.cpp`.
In `empty_cpu()` you can see that the `requires_grad()` field from the option is just ignored. Which is the bug I guess.

**Comment by prajjwal1** [comment]
I had a hard look at your suggestion. 
I looked at `native_functions.yaml`. I saw function which requires a grad flag like randn. They're defined in a similar manner as `new_` ones (only difference being variants is set as a method). Not every function calls `at::empty()`. For instance, `Tensor new_full` calls `at::full()`, `Tensor new_zeros` calls `at::zeros()`. I compared both the implementations of `empty_cuda` and `empty_cpu`, there is no `requires_grad` flag in both of them. I'm not seeing `grad` flag being taken as an argument in `TensorFactories.cpp` or `native_functions.yaml`. From what I can see, options are passed as *kwargs from `native_functions.yaml` in `TensorFactories.cpp`. I didn't see `requires_grad` being passed anywhere. Can you please tell what am I missing ?

**Comment by albanD** [comment]
The `requires_grad` arg is part of the `option`. It is added in there in `generated/python_variable_methods.cpp`.
The problem I think is that the core functions in aten don't handle it. But no other function above in the stack does either.

@ezyang do you have an idea where this arg should be handled in the stack? It feels wrong to have to handle that in every single kernel in aten native.

**Comment by prajjwal1** [comment]
I guessed that the `requires_grad` is bundled in the options, but I'm not seeing the source where it is being handled from the source. I grepped the `ATen/native` and `tools/autograd` for `requires_grad` for `new_` functions, didn't find the source.

**Comment by albanD** [comment]
Yes that's the issue. It is never handled :D

**Comment by prajjwal1** [comment]
Can it be added in `native_functions.yaml`? From there it will be passed on to `TensorFactories.cpp`. I guess that's how passing of arguments is being handled here.

**Comment by albanD** [comment]
If you want to handle it in `TensorFactories.cpp`, you can simply extract it from the option that are already there and handle it properly.
The problems with this are:
- aten should be fairly autograd agnostic. So it is weird to have autograd specific code in there.
- You will need to add this code for every single function that should handle that: It's going to be a lot of copy pasting...

Another approach could be that the VariableType should handle this and we mark these functions such that the codegen handles this parameter that is autograd specific.

But there were discussions already about requires_grad and options and I don't know what was the final decision? cc @izdeby @ezyang

**Comment by prajjwal1** [comment]
I think the latter option would be better. Could you please provide some pointers about how to go about that? By VariableType, do you mean when the function is wrapped into `wrap` or `THPVariable_wrap`?

**Comment by ezyang** [comment]
I think this may require a bit of refactoring to resolve "properly".

----

First, let me resolve the confusion about where `requires_grad` "should" be handled. The best way to answer this question is to first understand how `torch.empty`, which DOES handle `requires_grad` correctly, respects `requires_grad`. My intuition for where it should happen is that when you call `torch.empty` it eventually gets to `VariableType::empty` which handles respecting `requires_grad=True`. But we can see from the code that this is not true:

```
// in torch/csrc/autograd/generated/VariableTypeEverything.cpp 
Tensor empty_memory_format(IntArrayRef size, const TensorOptions & options, c10::optional<MemoryFormat> memory_for
mat) {
  RECORD_FUNCTION("empty", std::vector<c10::IValue>({}), Node::peek_at_next_sequence_nr());
  auto options_ = TensorOptions(options);
  // ...
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::empty(size, options_, memory_format);
  })();
  auto result = std::move(tmp);
  // ...
  return result;
} 

```

What actually happens is that the binding code calls `torch::empty`:

```
// torch/csrc/autograd/generated/python_torch_functions.cpp
static PyObject * THPVariable_empty(PyObject* self_, PyObject* args, PyObject* kwargs)
{
  HANDLE_TH_ERRORS
  static PythonArgParser parser({
    "empty(IntArrayRef size, *, DimnameList? names, MemoryFormat? memory_format=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
    "empty(IntArrayRef size, *, MemoryFormat? memory_format=None, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
  }, /*traceable=*/true);

  ParsedArgs<8> parsed_args;
  auto _r = parser.parse(args, kwargs, parsed_args);
  if(_r.has_torch_function()) {
    return handle_torch_function(_r, args, kwargs, THPVariableFunctionsModule, "torch");
  }
  switch (_r.idx) {
    case 0: {
      // aten::empty.names(int[] size, *, Dimname[]? names, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=None) -> Tensor
      auto __names = _r.toDimnameListOptional(1);
      c10::optional<DimnameList> names = __names ? c10::make_optional(DimnameList(__names.value())) : c10::nullopt;
      const auto options = TensorOptions()
          .dtype(_r.scalartype(3))
          .device(_r.device(5))
          .layout(_r.layoutOptional(4))
          .requires_grad(_r.toBool(7))
          .pinned_memory(_r.toBool(6));
      torch::utils::maybe_initialize_cuda(options);
      auto dispatch_empty = [](IntArrayRef size, c10::optional<DimnameList> names, const TensorOptions & options, c10::optional<MemoryFormat> memory_format) -> Tensor {
        pybind11::gil_scoped_release no_gil;
        return torch::empty(size, names, options, memory_format);
      };
      return wrap(dispatch_empty(_r.intlist(0), names, options, _r.memoryformatOptional(2)));
    }

```

and `torch::empty` is what actually handles setting requires grad!

```
// torch/csrc/autograd/generated/variable_factories.h
inline at::Tensor empty(at::IntArrayRef size, c10::optional<at::DimnameList> names, const at::TensorOptions & options = {}, c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
  // ...
  at::Tensor tensor = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::empty(size, names, at::TensorOptions(options), memory_format);
  })();
  at::Tensor result =
    autograd::make_variable(std::move(tensor), /*requires_grad=*/options.requires_grad());
  // ...
  return result;
}
```

So the problem with `new_full` is that it's a method, not a factory, and so it doesn't call a factory implementation that handles respecting requires grad.

----

So what should be done? Here are some guiding principles I have in my head:

* Variable handling (e.g., requires grad handling) should be done in VariableType
* torch::empty and at::empty should be the same thing

So it seems to me that the root cause of our problems is that the VariableType implementation should be responsible for handling requires grad (not torch::!) and if you fix the codegen there to handle it correctly, you'll also fix `new_full` while you're at it. But this codegen might be pretty involved, since you're going to essentially delete the `variable_factories.h` header entirely!

**Comment by prajjwal1** [comment]
All of these files are generated. Where do we need to add VariableType for `new_` so that it's reflected? I'm new to code generation workflow. As a result, I didn't get some bits clearly, could you please provide some pointers on how do I add this functionality and test it.  
Also, same problem exists for `new_empty` as `new_full` but `new_empty` calls `at::empty`. I highlighted it [here](https://github.com/pytorch/pytorch/issues/36455#issuecomment-613241396)

**Comment by albanD** [comment]
All these files are generated from the scripts in `tools/autograd/*`.
It is based on `derivatives.yaml` for the autograd part and the processed version of `native_functions.yaml` (in `aten/src/ATen/Declarations.yaml` IIRC).

**Comment by prajjwal1** [comment]
I had a look at `aten/src/ATen/Declarations.wrap`, it seems that `new_` functions are not defined there. I wanted to know where can I make the changes so they are reflected in `generated` directory.

**Comment by ezyang** [comment]
VariableType is specifically generated by `tools/autograd/gen_variable_type.py`. Declarations.cwrap shouldn't affect this at all.

I think fixing this may be quite complex and it may be difficult for us to walk you through it (we ourselves aren't exactly sure what changes are necessary). Take a look at that Python script and if it seems inscrutable it is probably best to pause on this issue until someone who is more familiar with the script can take a look at this.

**Comment by kurtamohler** [comment]
@prajjwal1, are you still planning on investigating more? I'm interested in taking a crack at it if not.

**Comment by prajjwal1** [comment]
> @prajjwal1, are you still planning on investigating more? I'm interested in taking a crack at it if not.

No, I'm not looking at this issue. @ezyang suggested that it would be better if someone more experienced would take a look. Feel free to try.

**Comment by DarthAnakin** [comment]
Is there anyway to fix this arg? Or has this been handled in the later version?

**Comment by ezyang** [comment]
should be fixed, please reopen if it's not

**Comment by nmll** [comment]
what's the follow-up?

**Comment by kurtamohler** [comment]
@nmll, this issue was fixed a while ago. If you're using a version of pytorch that is more than a year old, please update. Version 1.10 is the latest. If you are using an up to date version and you still see a problem, please submit a new issue
